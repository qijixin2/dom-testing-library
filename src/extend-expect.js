import extensions from './jest-extensions'

const {toBeInTheDOM, toHaveTextContent, toHaveAttribute} = extensions
expect.extend({toBeInTheDOM, toHaveTextContent, toHaveAttribute})
