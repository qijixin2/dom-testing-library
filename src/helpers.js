const globalObj = typeof window === 'undefined' ? global : window
// Constant node.nodeType for text nodes, see:
// https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants
const TEXT_NODE = 3

// Currently this fn only supports jest timers, but it could support other test runners in the future.
function runWithRealTimers(callback) {
  return _runWithRealTimers(callback).callbackReturnValue
}

function _runWithRealTimers(callback) {
  const timerAPI = {
    clearImmediate,
    clearInterval,
    clearTimeout,
    setImmediate,
    setInterval,
    setTimeout,
  }

  // istanbul ignore else
  if (typeof jest !== 'undefined') {
    jest.useRealTimers()
  }

  const callbackReturnValue = callback()

  const usedJestFakeTimers = Object.entries(timerAPI).some(
    ([name, func]) => func !== globalObj[name],
  )

  if (usedJestFakeTimers) {
    jest.useFakeTimers(timerAPI.setTimeout?.clock ? 'modern' : 'legacy')
  }

  return {
    callbackReturnValue,
    usedJestFakeTimers,
  }
}

const jestFakeTimersAreEnabled = () =>
  Boolean(_runWithRealTimers(() => {}).usedJestFakeTimers)

// we only run our tests in node, and setImmediate is supported in node.
// istanbul ignore next
function setImmediatePolyfill(fn) {
  return globalObj.setTimeout(fn, 0)
}

function getTimeFunctions() {
  // istanbul ignore next
  return {
    clearTimeoutFn: globalObj.clearTimeout,
    setImmediateFn: globalObj.setImmediate || setImmediatePolyfill,
    setTimeoutFn: globalObj.setTimeout,
  }
}

const {clearTimeoutFn, setImmediateFn, setTimeoutFn} = runWithRealTimers(
  getTimeFunctions,
)

function getDocument() {
  /* istanbul ignore if */
  if (typeof window === 'undefined') {
    throw new Error('Could not find default container')
  }
  return window.document
}
function getWindowFromNode(node) {
  if (node.defaultView) {
    // node is document
    return node.defaultView
  } else if (node.ownerDocument && node.ownerDocument.defaultView) {
    // node is a DOM node
    return node.ownerDocument.defaultView
  } else if (node.window) {
    // node is window
    return node.window
  } else if (node.then instanceof Function) {
    throw new Error(
      `It looks like you passed a Promise object instead of a DOM node. Did you do something like \`fireEvent.click(screen.findBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`, or await the findBy query \`fireEvent.click(await screen.findBy...\`?`,
    )
  } else {
    // The user passed something unusual to a calling function
    throw new Error(
      `Unable to find the "window" object for the given node. Please file an issue with the code that's causing you to see this error: https://github.com/testing-library/dom-testing-library/issues/new`,
    )
  }
}

function checkContainerType(container) {
  if (
    !container ||
    !(typeof container.querySelector === 'function') ||
    !(typeof container.querySelectorAll === 'function')
  ) {
    throw new TypeError(
      `Expected container to be an Element, a Document or a DocumentFragment but got ${getTypeName(
        container,
      )}.`,
    )
  }

  function getTypeName(object) {
    if (typeof object === 'object') {
      return object === null ? 'null' : object.constructor.name
    }
    return typeof object
  }
}

export {
  getWindowFromNode,
  getDocument,
  clearTimeoutFn as clearTimeout,
  setImmediateFn as setImmediate,
  setTimeoutFn as setTimeout,
  runWithRealTimers,
  checkContainerType,
  jestFakeTimersAreEnabled,
  TEXT_NODE,
}
