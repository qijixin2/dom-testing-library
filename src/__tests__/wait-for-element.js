import {waitForElement, wait} from '../'
// adds special assertions like toBeInTheDOM
import 'jest-dom/extend-expect'
import {render} from './helpers/test-utils'

async function skipSomeTime(delayMs) {
  await new Promise(resolve => setTimeout(resolve, delayMs))
}

async function skipSomeTimeForMutationObserver(delayMs = 50) {
  // Using `setTimeout` >30ms instead of `wait` here because `mutationobserver-shim` uses `setTimeout` ~30ms.
  await skipSomeTime(delayMs, 50)
}

test('it waits for the callback to return a value and only reacts to DOM mutations', async () => {
  const {container, getByTestId} = render(
    `<div data-testid="initial-element"></div>`,
  )

  let nextElIndex = 0
  const makeMutationFn = () => () => {
    container.appendChild(
      render(
        `<div data-testid="another-element-that-causes-mutation-${++nextElIndex}"></div>`,
      ).container.firstChild,
    )
  }

  const testEl = render(
    `<div data-testid="the-element-we-are-looking-for"></div>`,
  ).container.firstChild
  testEl.parentNode.removeChild(testEl)

  const mutationsAndCallbacks = [
    [
      makeMutationFn(),
      () => {
        throw new Error('First exception.')
      },
    ],
    [
      makeMutationFn(),
      () => {
        throw new Error('Second exception.')
      },
    ],
    [makeMutationFn(), () => null],
    [makeMutationFn(), () => undefined],
    [makeMutationFn(), () => getByTestId('the-element-we-are-looking-for')],
    [
      () => container.appendChild(testEl),
      () => getByTestId('the-element-we-are-looking-for'),
    ],
  ]

  const callback = jest
    .fn(() => {
      throw new Error('No more calls are expected.')
    })
    .mockName('callback')
    .mockImplementation(() => {
      throw new Error(
        'First callback call is synchronous, not returning any elements.',
      )
    })
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForElement(callback, {container}).then(
    successHandler,
    errorHandler,
  )

  // One synchronous `callback` call is expected.
  expect(callback).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  await skipSomeTimeForMutationObserver()

  // No more expected calls without DOM mutations.
  expect(callback).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  // Perform mutations one by one, waiting for each to trigger `MutationObserver`.
  for (const [mutationImpl, callbackImpl] of mutationsAndCallbacks) {
    callback.mockImplementation(callbackImpl)
    mutationImpl()
    await skipSomeTimeForMutationObserver() // eslint-disable-line no-await-in-loop
  }

  expect(callback).toHaveBeenCalledTimes(1 + mutationsAndCallbacks.length)
  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledWith(testEl)
  expect(errorHandler).toHaveBeenCalledTimes(0)
  expect(container).toMatchSnapshot()
  expect(testEl.parentNode).toBe(container)

  return promise
})

test('it waits for the next DOM mutation with default callback', async () => {
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForElement().then(successHandler, errorHandler)

  // Promise callbacks are always asynchronous.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  await skipSomeTimeForMutationObserver()

  // No more expected calls without DOM mutations.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  document.body.appendChild(document.createElement('div'))
  expect(document.body).toMatchSnapshot()

  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledWith(undefined)
  expect(errorHandler).toHaveBeenCalledTimes(0)
  expect(document.body).toMatchSnapshot()

  return promise
})

test('it waits characterData mutation', async () => {
  const {container} = render(`<div>initial text</div>`)

  const callback = jest
    .fn(() => container.textContent === 'new text')
    .mockName('callback')
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForElement(callback, {container}).then(
    successHandler,
    errorHandler,
  )

  // Promise callbacks are always asynchronous.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)
  expect(callback).toHaveBeenCalledTimes(1)
  expect(container).toMatchSnapshot()

  await skipSomeTimeForMutationObserver()

  expect(callback).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.querySelector('div').innerHTML = 'new text'
  await skipSomeTimeForMutationObserver()

  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler).toHaveBeenCalledTimes(0)
  expect(callback).toHaveBeenCalledTimes(2)
  expect(container).toMatchSnapshot()

  return promise
})

test('it waits for the attributes mutation', async () => {
  const {container} = render(``)

  const callback = jest
    .fn(() => container.getAttribute('data-test-attribute'))
    .mockName('callback')
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForElement(callback, {
    container,
  }).then(successHandler, errorHandler)

  expect(callback).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  await skipSomeTimeForMutationObserver()

  expect(callback).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.setAttribute('data-test-attribute', 'PASSED')
  await skipSomeTimeForMutationObserver()

  expect(callback).toHaveBeenCalledTimes(2)
  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledWith('PASSED')
  expect(errorHandler).toHaveBeenCalledTimes(0)

  return promise
})

test('it throws if timeout is exceeded', async () => {
  const {container} = render(``)

  const callback = jest.fn(() => null).mockName('callback')
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForElement(callback, {
    container,
    timeout: 70,
    mutationObserverOptions: {attributes: true},
  }).then(successHandler, errorHandler)

  expect(callback).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.setAttribute('data-test-attribute', 'something changed once')
  await skipSomeTimeForMutationObserver(50)

  expect(callback).toHaveBeenCalledTimes(2)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.setAttribute('data-test-attribute', 'something changed twice')
  await skipSomeTimeForMutationObserver(50)

  expect(callback).toHaveBeenCalledTimes(3)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler.mock.calls[0]).toMatchSnapshot()
  expect(container).toMatchSnapshot()

  return promise
})

test('it throws the same error that the callback has thrown if timeout is exceeded', async () => {
  const {container, getByTestId} = render(``)

  const callback = jest.fn(() => getByTestId('test')).mockName('callback')
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForElement(callback, {
    container,
    timeout: 70,
    mutationObserverOptions: {attributes: true},
  }).then(successHandler, errorHandler)

  expect(callback).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.setAttribute('data-test-attribute', 'something changed once')
  await skipSomeTimeForMutationObserver(50)

  expect(callback).toHaveBeenCalledTimes(2)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.setAttribute('data-test-attribute', 'something changed twice')
  await skipSomeTimeForMutationObserver(50)

  expect(callback).toHaveBeenCalledTimes(3)
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler.mock.calls[0]).toMatchSnapshot()
  expect(container).toMatchSnapshot()

  return promise
})

test('it returns immediately if the callback returns the value before any mutations', async () => {
  const {container, getByTestId} = render(
    `<div data-testid="initial-element"></div>`,
  )

  const callback = jest
    .fn(() => getByTestId('initial-element'))
    .mockName('callback')
  const successHandler = jest.fn().mockName('successHandler')
  const errorHandler = jest.fn().mockName('errorHandler')

  const promise = waitForElement(callback, {
    container,
    timeout: 70,
    mutationObserverOptions: {attributes: true},
  }).then(successHandler, errorHandler)

  // One synchronous `callback` call is expected.
  expect(callback).toHaveBeenCalledTimes(1)

  // The promise callbacks are expected to be called asyncronously.
  expect(successHandler).toHaveBeenCalledTimes(0)
  expect(errorHandler).toHaveBeenCalledTimes(0)
  await wait()
  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledWith(container.firstChild)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  container.setAttribute('data-test-attribute', 'something changed once')
  await skipSomeTimeForMutationObserver(50)

  // No more calls are expected.
  expect(callback).toHaveBeenCalledTimes(1)
  expect(successHandler).toHaveBeenCalledTimes(1)
  expect(errorHandler).toHaveBeenCalledTimes(0)

  expect(container).toMatchSnapshot()

  return promise
})
