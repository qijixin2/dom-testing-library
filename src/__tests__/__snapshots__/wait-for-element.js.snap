// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it returns immediately if the callback returns the value before any mutations 1`] = `
<div
  data-test-attribute="something changed once"
>
  <div
    data-testid="initial-element"
  />
</div>
`;

exports[`it throws if timeout is exceeded 1`] = `
Array [
  [Error: Timed out in waitForElement.],
]
`;

exports[`it throws if timeout is exceeded 2`] = `
<div
  data-test-attribute="something changed twice"
/>
`;

exports[`it throws the same error that the callback has thrown if timeout is exceeded 1`] = `
Array [
  [Error: Unable to find an element by: [data-testid="test"]

[36m<div[39m
  [33mdata-test-attribute[39m=[32m"something changed twice"[39m
[36m/>[39m],
]
`;

exports[`it throws the same error that the callback has thrown if timeout is exceeded 2`] = `
<div
  data-test-attribute="something changed twice"
/>
`;

exports[`it waits characterData mutation 1`] = `
<div>
  <div>
    initial text
  </div>
</div>
`;

exports[`it waits characterData mutation 2`] = `
<div>
  <div>
    new text
  </div>
</div>
`;

exports[`it waits for the callback to return a value and only reacts to DOM mutations 1`] = `
<div>
  <div
    data-testid="initial-element"
  />
  <div
    data-testid="another-element-that-causes-mutation-1"
  />
  <div
    data-testid="another-element-that-causes-mutation-2"
  />
  <div
    data-testid="another-element-that-causes-mutation-3"
  />
  <div
    data-testid="another-element-that-causes-mutation-4"
  />
  <div
    data-testid="another-element-that-causes-mutation-5"
  />
  <div
    data-testid="the-element-we-are-looking-for"
  />
</div>
`;

